First branch

atrishinnew opened the pull request:
pull request

atrishinnew committed:
Коммит на ридми.мд

atrishinnew committed:
Коммит на ридми.мд

atrishinnew committed:
Delete src directory

atrishinnew committed:
Коммит на ридми.мд

atrishinnew committed:
Merge remote-tracking branch 'origin/firstBranch' into firstBranch

# Conflicts:
#	src/Human.java

atrishinnew committed:
Коммит класса Book

atrishinnew committed:
Коммит класса Movie

atrishinnew committed:
Коммит класса Movie

atrishinnew committed:
Коммит класса Matrix

atrishinnew committed:
Коммит класса Calc (изменил только текст коммита для проверки)

KozlovDanila commented on src/Book.java:
```diff
@@ -0,0 +1,42 @@
+        this.creator = creator;
+    }
+
+    Book() {
```
Обычно пустой конструктор описывается первым, а затем уже остальные по нарастанию аргументов

atrishinnew replied:
Принял. Вроде действительно разумно и наверное удобно в будущем.

atrishinnew replied:
Изменил порядок следования описания конструкторов

KozlovDanila commented on src/Calc.java:
```diff
@@ -0,0 +1,43 @@
+
+
+    public static void  main (String [] args) {
+        int num1;
```
Нет смысла выносить переменную так отдельно. 
Сразу лучше писать int num1 = console.nextInt();

atrishinnew replied:
Понял. Вечно в голове вопрос, как правильно писать код, чтобы потом читать его было проще (наверное на эту тему тоже стоит что-то почитать).

KozlovDanila replied:
Это важная тема, но не на данном этапе. Думаю, не стоит сильно много сил на эту тему сейчас тратить. Ты будешь писать код и я тебе периодически буду говорить, что хорошо, а что плохо. И основные концепции с практикой запомнишь. А если хочется узнать больше, то всегда можно прочитать "чистый код" или "совершенный код". О них я тебе уже писал

KozlovDanila commented on src/Calc.java:
```diff
@@ -0,0 +1,43 @@
+        System.out.println(String.format("Результат деления введенных чисел (без остатка, ну его нахрен): " + calulate.division(num1, num2)));
+    }
+}
+//или я хрень какую то сделал или я не понял пока, для чего такие сложности. надеюсь это для того, чтобы просто набить руку
```
Нет, ты все сделал верно. Почему сложности? Все очень даже просто вроде. И да, это чисто, чтобы набить руку

KozlovDanila commented on src/Matrix.java:
```diff
@@ -0,0 +1,129 @@
+import java.util.Arrays;
+
+public class Matrix {
+    int[][] massiv = new int[4][4];
```
Почему массив сразу же каким-то образом проинициализирован? А если захочется матрицу не 4x4, а 5х5 например. Что тогда делать? Лучше массив передавать в конструкторе. Ну и название massiv глаза режет. Лучше дать другое название)

atrishinnew replied:
С учетом ответа в конце надо будет переосмыслить, похоже у меня даже мысли не возникло, что и это можно так завернуть. Подумаю над этим побольше.

KozlovDanila commented on src/Matrix.java:
```diff
@@ -0,0 +1,129 @@
+    int[][] massiv = new int[4][4];
+
+    //получить массив
+    void getMassiv(int[][] massiv) {
```
Метод get всегда подразумевает, что он будет что-то возвращать. Если ты пишешь getMassiv, значит и должен вернуться массив.

atrishinnew replied:
Тут как раз у меня мозг боролся, или вывести на экран сразу в методе (нафиг гонять лишние переходы) или вернуть из метода матрицу/массив и потом уже в main его вывести в консоль. 
Учту, что раз get, значит надо возвращать.

KozlovDanila commented on src/Matrix.java:
```diff
@@ -0,0 +1,129 @@
+
+    }
+
+}
```
Класс матрицы создал, а для чего создал - не понятно, ибо все равно все действия выполняешь так, будто не с классом работаешь, а с кусками кода. Но это на первых порах нормально. 
Класс - это представление какой-то конкретной сущности. В данном случае матрицы. Если представить матрицу, то она может быть какой угодно по размеру. 5x5, 10x10, хоть 100х100. Соответственно, класс матрицы должен в себе лишь описывать массив, но не должен говорить сразу же какого размера этот массив будет. Ведь класс - это только описание, некая инструкция. Указывать же конкретную размерность надо в момент создания объекта. 
```
int[][] array = new int[n][m];
// заполнение, либо иная работа
Matrix matrix = new Matrix(array);
и пошёл работать с матрицей.
```

Соответственное, когда думаешь о классе и его объекте, как о конкретной сущности, то сразу возникают вопросы:
1. Метод "вернуть массив" что должен возвращать? Ответ: сам массив, то есть полное содержимое матрицы в виде двумерного массива.
2. Метод "получить значение по индексу" что должен в себя принимать и что должен возвращать? Ответ: принимать индексы, возвращать конкретное значение по этим индексам.
3. Метод "умножить МАТРИЦЫ" что должен в себя принимать и возвращать? Когда мы говорим "одна матрица умножается на другую", мы говорим слово МАТРИЦА. Значит на вход тоже должна быть матрица. После результата умножения мы что получаем? Новую матрицу. Значит и на выход должна быть новая матрица.
4. Другие методы в том же стиле

И когда мы это все опишем, то у нас будет конкретный класс, который для использования будет именно МАТРИЦЕЙ,  а не набором методов для работы с массивами

atrishinnew replied:
Переосмыслю.
В этих примерах у меня просто в голове не укладывалось, зачем мне вызывать какой-то отдельный метод для вывода в консоль, или перемножения, в общем для простых действий, когда проще сразу в main взять и сделать это.
Понимаю, что в дальнейшем, в этих методах будет много логики и не просто вывод в консоль, но пока голова это все осознает, сложно скрестить обычную человеческую логику с логикой из примера.
Тут возникали мысли, что возможно само задание сделать для "таких", как я)

KozlovDanila replied:
А это как раз один из правильных подходов к написанию кода. Есть одно из правил, что класс должен выполнять только одно действие. Если класс является матрицей и при этом сам себя печатает, то он дейлает два действия. Соответственно, нужно это разделить. Матрица, как единая сущность, делает только то, что должна делать, а печатает какая-то другая сущность, которая отвечает за то, чтобы печатать. Чем это полезно? Тем, что это можно переиспользовать. Класс, ответственный за печать, может напечатать что угодно. А матрица может быть полезна и без печати. Плюс куски кода при таком разделении обязаностей становятся цельными без дополнительной логики, которая не относится к тому, что код делает.

atrishinnew committed:
Изменил порядок описания конструкторов (пустой поставил первым).

atrishinnew committed:
-=Новая матрица=-
1. Сделал статические переменные и статический метод для того, чтобы задать размерность матрицы
2. создал конструктор с полями и создаю новый объект с матрицей, у которой размерность задал в п.1
3. создал метод заполнения матрицы на вход которому ничего не передаю. Метод заполнит матрицу для объекта из п.2

4. а потом вывод всего дела в консоль

atrishinnew committed:
1. создал конструктор с полем array
2. вызвал метод для задания количества строк и столбцов
3. создал локальный массив в main
4. созданный массив отдал на вход, при создании объекта firstMatrix
5. вызвал метод firstMatrix.setMatrix() для заполнения матрицы мусором

atrishinnew committed:
1. заполнил массив сразу в main
2. передал этот массив на вход конструктору
3. в таком варианте понятно, почему содержимое массива array1 равно содержимомум массива в объекте irstMatrix.array

Но по большому счету, я не понимаю, в чем принципиальная разница. Ведь работает )))

atrishinnew committed:
Переписал пока без тестов.
надеюсь стало лучше )

KozlovDanila commented on src/Matrix02.java:
```diff
@@ -1,64 +1,162 @@
+    }
+
+    //сложение массивов
+    int[][] sumMatrix (int[][] matrix1, int[][] matrix2) {
```
Все равно на вход именно массив(аж два), и на выход массив. А хотелось бы на вход матрицу(одну) и на выход матрицу)

KozlovDanila commented on src/Matrix02.java:
```diff
@@ -1,64 +1,162 @@
+        fifthM.multiMatrix(secondM.array,secondM.array);
+        System.out.println("=====");
+        System.out.println(Arrays.deepToString(fifthM.array));
+
```
Ну и нет смысла писать аткие тесты, ибо это не тесты. Если хочется писать тесты, нужно использовать junit

atrishinnew committed:
Программа, которая подсчитывает кол-во пробелов в фразе, которую введет пользователь.
Написал тест.

atrishinnew committed:
Программа, которая подсчитывает кол-во пробелов в фразе, которую введет пользователь.
Написал тест.
(второй вариант)
